source/common/config/delta_subscription_state.h

```diff
@@ -80,36 +85,37 @@ class DeltaSubscriptionState : public Logger::Loggable<Logger::Id::config> {
     absl::optional<std::string> version_;
   };
 
-  // Use these helpers to ensure resource_state_ and resource_names_ get updated together.
-  void addResourceState(const envoy::service::discovery::v3::Resource& resource);
-  void setResourceWaitingForServer(const std::string& resource_name);
-  void removeResourceState(const std::string& resource_name);
+  void addResourceStateFromServer(const envoy::service::discovery::v3::Resource& resource);
+  OptRef<ResourceState> getRequestedResourceState(absl::string_view resource_name);
+  OptRef<const ResourceState> getRequestedResourceState(absl::string_view resource_name) const;
 
-  void populateDiscoveryRequest(envoy::service::discovery::v3::DeltaDiscoveryResponse& request);
+  bool isInitialRequestForLegacyWildcard();
 
   // A map from resource name to per-resource version. The keys of this map are exactly the resource
   // names we are currently interested in. Those in the waitingForServer state currently don't have
   // any version for that resource: we need to inform the server if we lose interest in them, but we
   // also need to *not* include them in the initial_resource_versions map upon a reconnect.
-  absl::node_hash_map<std::string, ResourceState> resource_state_;
+  absl::node_hash_map<std::string, ResourceState> requested_resource_state_;
+  // A map from resource name to per-resource version. The keys of this map are resource names we
+  // have received as a part of the wildcard subscription.
+  absl::node_hash_map<std::string, std::string> wildcard_resource_state_;
+  // Used for storing resources that we lost interest in, but could
+  // also be a part of wildcard subscription.
+  absl::node_hash_map<std::string, std::string> ambiguous_resource_state_;
```

Comment:
Just getting back to review here (sorry for delay). I was a little surprised by the `ambiguous_resource_state_` and then I read your explanation and it made sense. Do you think you could document the resource state machine here or somewhere else for posterity?

Reply:
I wrote a description of the state machine in `DeltaSubscriptionState` class docs. Please let me know if they are any good. I also added two pictures. I wasn't sure where to keep them, so currently they are also in `source/common/config` directory. Pasting them here inline too:

non wildcard resource state machine:
![non-wildcard-resource-state-machine](https://user-images.githubusercontent.com/96081/131704593-1cd0fb3f-f3b5-4655-b822-81623e1e8fd4.png)

wildcard resource state machine:
![wildcard-resource-state-machine](https://user-images.githubusercontent.com/96081/131704652-63d27861-742d-479c-8222-f9d093133c93.png)

Reply:
These diagrams look great!

Reply:
Thanks. :)

Reply:
+1, thanks so much.